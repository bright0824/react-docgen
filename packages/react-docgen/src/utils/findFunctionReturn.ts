import type { NodePath } from '@babel/traverse';
import { visitors } from '@babel/traverse';
import resolveToValue from './resolveToValue';
import { ignore } from './traverse';

type Predicate = (p: NodePath) => boolean;
interface TraverseState {
  readonly predicate: Predicate;
  resolvedReturnPath?: NodePath;
  readonly seen: WeakSet<NodePath>;
}

const explodedVisitors = visitors.explode<TraverseState>({
  Function: { enter: ignore },
  Class: { enter: ignore },
  ObjectExpression: { enter: ignore },
  ReturnStatement: {
    enter: function (path, state) {
      const argument = path.get('argument');

      if (argument.hasNode()) {
        const resolvedPath = resolvesToFinalValue(
          argument,
          state.predicate,
          state.seen,
        );

        if (resolvedPath) {
          state.resolvedReturnPath = resolvedPath;
          path.stop();
        }
      }
    },
  },
});

function resolvesToFinalValue(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  // avoid returns with recursive function calls
  if (seen.has(path)) {
    return;
  }
  seen.add(path);

  // Is the path already passes then return it.
  if (predicate(path)) {
    return path;
  }

  const resolvedPath = resolveToValue(path);

  // If the resolved path is already passing then no need to further check
  // Only do this if the resolvedPath actually resolved something as otherwise we did this check already
  if (resolvedPath.node !== path.node && predicate(resolvedPath)) {
    return resolvedPath;
  }

  // If the path points to a conditional expression, then we need to look only at
  // the two possible paths
  if (resolvedPath.isConditionalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('consequent'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('alternate'), predicate, seen)
    );
  }

  // If the path points to a logical expression (AND, OR, ...), then we need to look only at
  // the two possible paths
  if (resolvedPath.isLogicalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('left'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('right'), predicate, seen)
    );
  }

  // If we have a call expression, lets try to follow it
  if (resolvedPath.isCallExpression()) {
    const returnValue = findFunctionReturnWithCache(
      resolveToValue(resolvedPath.get('callee')),
      predicate,
      seen,
    );

    if (returnValue) {
      return returnValue;
    }
  }

  return;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
function findFunctionReturnWithCache(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  let functionPath: NodePath = path;

  if (functionPath.isObjectProperty()) {
    functionPath = functionPath.get('value');
  } else if (functionPath.isClassProperty()) {
    const classPropertyValue = functionPath.get('value');

    if (classPropertyValue.hasNode()) {
      functionPath = classPropertyValue;
    }
  }

  if (!functionPath.isFunction()) {
    return;
  }

  // skip traversing for ArrowFunctionExpressions with no block
  if (path.isArrowFunctionExpression()) {
    const body = path.get('body');

    if (!body.isBlockStatement()) {
      return resolvesToFinalValue(body, predicate, seen);
    }
  }

  const state: TraverseState = {
    predicate,
    seen,
  };

  path.traverse(explodedVisitors, state);

  return state.resolvedReturnPath;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
export default function findFunctionReturn(
  path: NodePath,
  predicate: Predicate,
): NodePath | undefined {
  return findFunctionReturnWithCache(path, predicate, new WeakSet());
}
