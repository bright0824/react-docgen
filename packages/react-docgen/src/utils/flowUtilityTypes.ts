import type { NodePath } from '@babel/traverse';
import type { GenericTypeAnnotation } from '@babel/types';

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(
  path: NodePath,
): path is NodePath<GenericTypeAnnotation> {
  if (path.isGenericTypeAnnotation()) {
    const idPath = path.get('id');

    if (idPath.isIdentifier()) {
      const name = idPath.node.name;

      return name === '$Exact' || name === '$ReadOnly';
    }
  }

  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  let resultPath: NodePath = path;
  while (isSupportedUtilityType(resultPath)) {
    const typeParameters = resultPath.get('typeParameters');
    if (!typeParameters.hasNode()) break;

    const firstParam = typeParameters.get('params')[0];
    if (!firstParam) break;

    resultPath = firstParam;
  }

  return resultPath;
}
