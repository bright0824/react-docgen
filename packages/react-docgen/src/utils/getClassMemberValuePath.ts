import type { NodePath } from '@babel/traverse';
import type {
  ClassDeclaration,
  ClassExpression,
  ClassMethod,
  ClassProperty,
  Expression,
} from '@babel/types';
import getNameOrValue from './getNameOrValue';

export default function getClassMemberValuePath(
  classDefinition: NodePath<ClassDeclaration | ClassExpression>,
  memberName: string,
): NodePath<ClassMethod | Expression> | null {
  const classMember = classDefinition
    .get('body')
    .get('body')
    .find(memberPath => {
      if (
        (memberPath.isClassMethod() && memberPath.node.kind !== 'set') ||
        memberPath.isClassProperty()
      ) {
        const key = (memberPath as NodePath<ClassMethod | ClassProperty>).get(
          'key',
        );

        return (
          (!memberPath.node.computed || key.isLiteral()) &&
          getNameOrValue(key) === memberName
        );
      }

      return false;
    });

  if (classMember) {
    // For ClassProperty we return the value and for ClassMethod
    // we return itself
    return classMember.isClassMethod()
      ? classMember
      : (classMember.get('value') as NodePath<Expression>);
  }

  return null;
}
