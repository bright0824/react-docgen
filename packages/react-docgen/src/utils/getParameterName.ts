import type { NodePath } from '@babel/traverse';
import type {
  ArrayPattern,
  AssignmentPattern,
  Identifier,
  ObjectPattern,
  RestElement,
  TSParameterProperty,
} from '@babel/types';
import printValue from './printValue.js';

type ParameterNodePath = NodePath<
  | ArrayPattern
  | AssignmentPattern
  | Identifier
  | ObjectPattern
  | RestElement
  | TSParameterProperty
>;

export default function getParameterName(
  parameterPath: ParameterNodePath,
): string {
  if (parameterPath.isIdentifier()) {
    return parameterPath.node.name;
  } else if (parameterPath.isAssignmentPattern()) {
    return getParameterName(parameterPath.get('left') as ParameterNodePath);
  } else if (
    parameterPath.isObjectPattern() ||
    parameterPath.isArrayPattern()
  ) {
    return printValue(parameterPath);
  } else if (parameterPath.isRestElement()) {
    return `...${getParameterName(
      parameterPath.get('argument') as ParameterNodePath,
    )}`;
  } else if (parameterPath.isTSParameterProperty()) {
    return getParameterName(parameterPath.get('parameter'));
  }

  throw new TypeError(
    'Parameter name must be one of Identifier, AssignmentPattern, ArrayPattern, ' +
      `ObjectPattern or RestElement, instead got ${parameterPath.node.type}`,
  );
}
