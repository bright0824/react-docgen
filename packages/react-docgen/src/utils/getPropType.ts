import type { NodePath } from '@babel/traverse';
import { getDocblock } from '../utils/docblock.js';
import getMembers from './getMembers.js';
import getPropertyName from './getPropertyName.js';
import isRequiredPropType from '../utils/isRequiredPropType.js';
import printValue from './printValue.js';
import resolveToValue from './resolveToValue.js';
import resolveObjectKeysToArray from './resolveObjectKeysToArray.js';
import resolveObjectValuesToArray from './resolveObjectValuesToArray.js';
import type { PropTypeDescriptor } from '../Documentation.js';
import type { ArrayExpression, Expression, SpreadElement } from '@babel/types';

function getEnumValuesFromArrayExpression(
  path: NodePath<ArrayExpression>,
): Array<Record<string, unknown>> {
  const values: Array<Record<string, unknown>> = [];

  path.get('elements').forEach((elementPath) => {
    if (!elementPath.hasNode()) return;

    if (elementPath.isSpreadElement()) {
      const value = resolveToValue(elementPath.get('argument'));

      if (value.isArrayExpression()) {
        // if the SpreadElement resolved to an Array, add all their elements too
        return values.push(...getEnumValuesFromArrayExpression(value));
      } else {
        // otherwise we'll just print the SpreadElement itself
        return values.push({
          value: printValue(elementPath),
          computed: !elementPath.isLiteral(),
        });
      }
    }

    // try to resolve the array element to it's value
    const value = resolveToValue(elementPath as NodePath<Expression>);

    return values.push({
      value: printValue(value.isImportDeclaration() ? elementPath : value),
      computed: !value.isLiteral(),
    });
  });

  return values;
}

function getPropTypeOneOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const value = resolveToValue(argumentPath);

  if (value.isArrayExpression()) {
    type.value = getEnumValuesFromArrayExpression(value);
  } else {
    const objectValues =
      resolveObjectKeysToArray(value) || resolveObjectValuesToArray(value);

    if (objectValues) {
      type.value = objectValues.map((objectValue) => ({
        value: objectValue,
        computed: false,
      }));
    } else {
      // could not easily resolve to an Array, let's print the original value
      type.computed = true;
      type.value = printValue(argumentPath);
    }
  }

  return type;
}

function getPropTypeOneOfType(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  if (argumentPath.isArrayExpression()) {
    type.value = argumentPath.get('elements').map((elementPath) => {
      if (!elementPath.hasNode()) return;
      const descriptor: PropTypeDescriptor = getPropType(elementPath);
      const docs = getDocblock(
        elementPath as NodePath<Expression | SpreadElement>,
      );

      if (docs) {
        descriptor.description = docs;
      }

      return descriptor;
    });
  }

  return type;
}

function getPropTypeArrayOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const docs = getDocblock(argumentPath);

  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath);

  // @ts-ignore
  if (subType.name !== 'unknown') {
    type.value = subType;
  }

  return type;
}

function getPropTypeObjectOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const docs = getDocblock(argumentPath);

  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath);

  // @ts-ignore
  if (subType.name !== 'unknown') {
    type.value = subType;
  }

  return type;
}

function getFirstArgument(path: NodePath): NodePath | undefined {
  let argument: NodePath | undefined;

  if (path.isCallExpression()) {
    argument = path.get('arguments')[0];
  } else {
    const members = getMembers(path, true);

    if (members[0] && members[0].argumentPaths[0]) {
      argument = members[0].argumentPaths[0];
    }
  }

  return argument;
}

function isCyclicReference(
  argument: NodePath,
  argumentPath: NodePath,
): boolean {
  return Boolean(argument && resolveToValue(argument) === argumentPath);
}

/**
 * Handles shape and exact prop types
 */
function getPropTypeShapish(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  if (!argumentPath.isObjectExpression()) {
    argumentPath = resolveToValue(argumentPath);
  }

  if (argumentPath.isObjectExpression()) {
    const value: Record<string, PropTypeDescriptor> = {};
    let rawValue: string | undefined;

    argumentPath.get('properties').forEach((propertyPath) => {
      // We only handle ObjectProperty as there is nothing to handle for
      // SpreadElements and ObjectMethods
      if (propertyPath.isObjectProperty()) {
        const propertyName = getPropertyName(propertyPath);

        if (!propertyName) return;

        const valuePath = propertyPath.get('value');
        const argument = getFirstArgument(valuePath);

        // This indicates we have a cyclic reference in the shape
        // In this case we simply print the argument to shape and bail
        if (argument && isCyclicReference(argument, argumentPath)) {
          rawValue = printValue(argument);

          return;
        }

        const descriptor = getPropType(valuePath);
        const docs = getDocblock(propertyPath);

        if (docs) {
          descriptor.description = docs;
        }
        descriptor.required = isRequiredPropType(valuePath);
        value[propertyName] = descriptor;
      }
    });

    type.value = rawValue ?? value;
  }

  return type;
}

function getPropTypeInstanceOf(
  _type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  return {
    name: 'instanceOf',
    value: printValue(argumentPath),
  };
}

const simplePropTypes = [
  'array',
  'bool',
  'func',
  'number',
  'object',
  'string',
  'any',
  'element',
  'node',
  'symbol',
  'elementType',
] as const;

function isSimplePropType(
  name: string,
): name is (typeof simplePropTypes)[number] {
  return simplePropTypes.includes(name as (typeof simplePropTypes)[number]);
}

type PropTypeHandler = (
  type: PropTypeDescriptor,
  argumentPath: NodePath,
) => PropTypeDescriptor;

const propTypes = new Map<
  string,
  (argumentPath: NodePath | undefined) => PropTypeDescriptor
>([
  ['oneOf', callPropTypeHandler.bind(null, 'enum', getPropTypeOneOf)],
  ['oneOfType', callPropTypeHandler.bind(null, 'union', getPropTypeOneOfType)],
  [
    'instanceOf',
    callPropTypeHandler.bind(null, 'instanceOf', getPropTypeInstanceOf),
  ],
  ['arrayOf', callPropTypeHandler.bind(null, 'arrayOf', getPropTypeArrayOf)],
  ['objectOf', callPropTypeHandler.bind(null, 'objectOf', getPropTypeObjectOf)],
  ['shape', callPropTypeHandler.bind(null, 'shape', getPropTypeShapish)],
  ['exact', callPropTypeHandler.bind(null, 'exact', getPropTypeShapish)],
]);

function callPropTypeHandler(
  name: PropTypeDescriptor['name'],
  handler: PropTypeHandler,
  argumentPath: NodePath | undefined,
): PropTypeDescriptor {
  let type: PropTypeDescriptor = { name };

  if (argumentPath) {
    type = handler(type, argumentPath);
  }

  if (!type.value) {
    // If there is no argument then leave the value an empty string
    type.value = argumentPath ? printValue(argumentPath) : '';
    type.computed = true;
  }

  return type;
}

/**
 * Tries to identify the prop type by inspecting the path for known
 * prop type names. This method doesn't check whether the found type is actually
 * from React.PropTypes. It simply assumes that a match has the same meaning
 * as the React.PropTypes one.
 *
 * If there is no match, "custom" is returned.
 */
export default function getPropType(path: NodePath): PropTypeDescriptor {
  let descriptor: PropTypeDescriptor | null = null;

  getMembers(path, true).some((member) => {
    const memberPath = member.path;
    let name: string | null = null;

    if (memberPath.isStringLiteral()) {
      name = memberPath.node.value;
    } else if (memberPath.isIdentifier() && !member.computed) {
      name = memberPath.node.name;
    }
    if (name) {
      if (isSimplePropType(name)) {
        descriptor = { name };

        return true;
      }
      const propTypeHandler = propTypes.get(name);

      if (propTypeHandler) {
        descriptor = propTypeHandler(member.argumentPaths[0]);

        return true;
      }
    }

    return;
  });

  if (descriptor) {
    return descriptor;
  }

  if (path.isIdentifier() && isSimplePropType(path.node.name)) {
    return { name: path.node.name };
  }

  if (path.isCallExpression()) {
    const callee = path.get('callee');

    if (callee.isIdentifier()) {
      const propTypeHandler = propTypes.get(callee.node.name);

      if (propTypeHandler) {
        return propTypeHandler(path.get('arguments')[0]);
      }
    }
  }

  return { name: 'custom', raw: printValue(path) };
}
