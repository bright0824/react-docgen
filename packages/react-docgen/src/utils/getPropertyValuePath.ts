import type { NodePath } from '@babel/traverse';
import type { Expression, ObjectExpression, ObjectMethod } from '@babel/types';
import getPropertyName from './getPropertyName.js';

/**
 * Given an ObjectExpression, this function returns the path of the value of
 * the property with name `propertyName`. if the property is an ObjectMethod we
 * return the ObjectMethod itself.
 */
export default function getPropertyValuePath(
  path: NodePath<ObjectExpression>,
  propertyName: string,
): NodePath<Expression | ObjectMethod> | null {
  const property = path
    .get('properties')
    .find(
      propertyPath =>
        !propertyPath.isSpreadElement() &&
        getPropertyName(propertyPath) === propertyName,
    );

  if (property) {
    return property.isObjectMethod()
      ? property
      : (property.get('value') as NodePath<Expression>);
  }

  return null;
}
