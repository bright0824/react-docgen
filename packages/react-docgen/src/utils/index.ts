import * as docblock from './docblock';
import * as expressionTo from './expressionTo';
import * as flowUtilityTypes from './flowUtilityTypes';
import * as traverse from './traverse';

export { docblock, expressionTo, flowUtilityTypes, traverse };

export { default as getClassMemberValuePath } from './getClassMemberValuePath';
export { default as getFlowType } from './getFlowType';
export { default as getMemberExpressionRoot } from './getMemberExpressionRoot';
export { default as getMemberExpressionValuePath } from './getMemberExpressionValuePath';
export { default as getMembers } from './getMembers';
export {
  default as getMemberValuePath,
  isSupportedDefinitionType,
} from './getMemberValuePath';
export { default as getMethodDocumentation } from './getMethodDocumentation';
export type { MethodNodePath } from './getMethodDocumentation';
export { default as getNameOrValue } from './getNameOrValue';
export { default as getParameterName } from './getParameterName';
export { default as getPropertyName, COMPUTED_PREFIX } from './getPropertyName';
export { default as getPropertyValuePath } from './getPropertyValuePath';
export { default as getPropType } from './getPropType';
export { default as getTSType } from './getTSType';
export { default as getTypeAnnotation } from './getTypeAnnotation';
export {
  default as getTypeFromReactComponent,
  applyToTypeProperties,
} from './getTypeFromReactComponent';
export { default as getTypeIdentifier } from './getTypeIdentifier';
export { default as getTypeParameters } from './getTypeParameters';
export type { TypeParameters } from './getTypeParameters';
export { default as isDestructuringAssignment } from './isDestructuringAssignment';
export { default as isExportsOrModuleAssignment } from './isExportsOrModuleAssignment';
export { default as isReactBuiltinCall } from './isReactBuiltinCall';
export { default as isReactChildrenElementCall } from './isReactChildrenElementCall';
export { default as isReactCloneElementCall } from './isReactCloneElementCall';
export { default as isReactComponentClass } from './isReactComponentClass';
export { default as isReactComponentMethod } from './isReactComponentMethod';
export { default as isReactCreateClassCall } from './isReactCreateClassCall';
export { default as isReactCreateElementCall } from './isReactCreateElementCall';
export { default as isReactForwardRefCall } from './isReactForwardRefCall';
export { default as isReactModuleName } from './isReactModuleName';
export { default as isRequiredPropType } from './isRequiredPropType';
export { default as isStatelessComponent } from './isStatelessComponent';
export { default as isUnreachableFlowType } from './isUnreachableFlowType';
export { default as match } from './match';
export { default as normalizeClassDefinition } from './normalizeClassDefinition';
export { default as parseJsDoc } from './parseJsDoc';
export { default as postProcessDocumentation } from './postProcessDocumentation';
export { default as printValue } from './printValue';
export { default as resolveExportDeclaration } from './resolveExportDeclaration';
export { default as resolveFunctionDefinitionToReturnValue } from './resolveFunctionDefinitionToReturnValue';
export { default as resolveGenericTypeAnnotation } from './resolveGenericTypeAnnotation';
export { default as resolveHOC } from './resolveHOC';
export { default as resolveObjectPatternPropertyToValue } from './resolveObjectPatternPropertyToValue';
export {
  default as resolveObjectKeysToArray,
  resolveObjectToNameArray,
} from './resolveObjectKeysToArray';
export { default as resolveObjectValuesToArray } from './resolveObjectValuesToArray';
export { default as resolveToModule } from './resolveToModule';
export { default as resolveToValue } from './resolveToValue';
export { default as setPropDescription } from './setPropDescription';
