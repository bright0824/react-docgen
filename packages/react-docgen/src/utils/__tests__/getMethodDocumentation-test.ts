import type { NodePath } from '@babel/traverse';
import type {
  AssignmentExpression,
  ClassDeclaration,
  ClassMethod,
  ClassPrivateMethod,
  ClassProperty,
  ExpressionStatement,
  ObjectExpression,
  ObjectMethod,
} from '@babel/types';
import { parse, makeMockImporter, parseTypescript } from '../../../tests/utils';
import getMethodDocumentation from '../getMethodDocumentation.js';
import { describe, expect, test } from 'vitest';

describe('getMethodDocumentation', () => {
  const mockImporter = makeMockImporter({
    hello: (stmt) =>
      stmt(`
      export default () => {};
    `).get('declaration'),

    bar: (stmt) =>
      stmt(`
      export default (bar: number) => {};
    `).get('declaration'),

    baz: (stmt) =>
      stmt(`
      export default (): number => {};
    `).get('declaration'),
  });

  describe('name', () => {
    test('extracts the method name', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          hello() {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

      expect(getMethodDocumentation(method)).toEqual({
        name: 'hello',
        docblock: null,
        modifiers: [],
        returns: null,
        params: [],
      });
    });

    test('handles function assignment', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          hello = () => {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassProperty>;

      expect(getMethodDocumentation(method)).toEqual({
        name: 'hello',
        docblock: null,
        modifiers: [],
        returns: null,
        params: [],
      });
    });

    test('handles computed method name', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          [foo]() {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

      expect(getMethodDocumentation(method)).toMatchSnapshot();
    });

    test('ignores complex computed method name', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          [() => {}]() {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

      expect(getMethodDocumentation(method)).toMatchSnapshot();
    });

    test('resolves assignment of imported function', () => {
      const def = parse.statement<ClassDeclaration>(
        `
        class Foo {
          hello = hello;
        }
        import hello from 'hello';
      `,
        mockImporter,
      );
      const method = def.get('body').get('body')[0] as NodePath<ClassProperty>;

      expect(getMethodDocumentation(method)).toEqual({
        name: 'hello',
        docblock: null,
        modifiers: [],
        returns: null,
        params: [],
      });
    });
  });

  describe('docblock', () => {
    test('extracts the method docblock', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          /**
           * Don't use this!
           */
          foo() {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

      expect(getMethodDocumentation(method)).toEqual({
        name: 'foo',
        docblock: "Don't use this!",
        modifiers: [],
        returns: null,
        params: [],
      });
    });

    test('extracts docblock on property assignment', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          /**
           * Don't use this!
           */
          foo = () => {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassProperty>;

      expect(getMethodDocumentation(method)).toEqual({
        name: 'foo',
        docblock: "Don't use this!",
        modifiers: [],
        returns: null,
        params: [],
      });
    });

    test('extracts docblock on function assignment', () => {
      const method = parse
        .statementLast<ExpressionStatement>(
          `const Foo = () => {}
           /**
            * Don't use this!
            */
           Foo.foo = () => {}
          `,
        )
        .get('expression') as NodePath<AssignmentExpression>;

      expect(getMethodDocumentation(method)).toEqual({
        name: 'foo',
        docblock: "Don't use this!",
        modifiers: ['static'],
        returns: null,
        params: [],
      });
    });
  });

  describe('parameters', () => {
    test('extracts flow type info', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          foo(bar: number) {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

      expect(getMethodDocumentation(method)).toMatchSnapshot();
    });

    test('extracts flow type info', () => {
      const def = parseTypescript.statement<ClassDeclaration>(`
        class Foo {
          foo(bar: number) {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

      expect(getMethodDocumentation(method)).toMatchSnapshot();
    });

    test('does not add type parameters to alias', () => {
      const def = parseTypescript.statement<ClassDeclaration>(`
        class Foo<T> {
          foo(bar: Foo<T>) {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

      expect(getMethodDocumentation(method)).toMatchSnapshot();
    });

    test('extracts flow type info on function assignment', () => {
      const def = parse.statement<ClassDeclaration>(`
        class Foo {
          foo = (bar: number) => {}
        }
      `);
      const method = def.get('body').get('body')[0] as NodePath<ClassProperty>;

      expect(getMethodDocumentation(method)).toMatchSnapshot();
    });

    test('resolves flow type info on imported functions', () => {
      const def = parse.statement<ClassDeclaration>(
        `
        class Foo {
          foo = bar
        }
        import bar from 'bar';
      `,
        mockImporter,
      );
      const method = def.get('body').get('body')[0] as NodePath<ClassProperty>;

      expect(getMethodDocumentation(method)).toMatchSnapshot();
    });

    describe('modifiers', () => {
      function methodModifiersDoc(modifiers) {
        return {
          name: 'foo',
          docblock: null,
          modifiers,
          returns: null,
          params: [],
        };
      }

      test('detects no modifiers', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            foo() {}
          }
        `);
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toEqual(methodModifiersDoc([]));
      });

      test('detects static functions', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            static foo() {}
          }
        `);
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toEqual(
          methodModifiersDoc(['static']),
        );
      });

      test('detects manually set static functions', () => {
        const def = parse.expression<ObjectExpression>(`{ foo() {} }`);
        const method = def.get('properties')[0] as NodePath<ObjectMethod>;

        expect(getMethodDocumentation(method, { isStatic: true })).toEqual(
          methodModifiersDoc(['static']),
        );
      });

      test('detects generators', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            *foo () {}
          }
        `);
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toEqual(
          methodModifiersDoc(['generator']),
        );
      });

      test('detects async functions', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            async foo () {}
          }
        `);
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toEqual(
          methodModifiersDoc(['async']),
        );
      });

      test('detects static async functions', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            static async foo () {}
          }
        `);
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toEqual(
          methodModifiersDoc(['static', 'async']),
        );
      });
    });

    describe('returns', () => {
      function methodReturnDoc(returnValue) {
        return {
          name: 'foo',
          docblock: null,
          modifiers: [],
          returns: returnValue,
          params: [],
        };
      }

      test('returns null if return is not documented', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            foo () {}
          }
        `);
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toEqual(methodReturnDoc(null));
      });

      test('extracts flow types', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            foo (): number {}
          }
        `);
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toEqual(
          methodReturnDoc({
            type: { name: 'number' },
          }),
        );
      });

      test('extracts flow types on function assignment', () => {
        const def = parse.statement<ClassDeclaration>(`
          class Foo {
            foo = (): number => {}
          }
        `);
        const method = def
          .get('body')
          .get('body')[0] as NodePath<ClassProperty>;

        expect(getMethodDocumentation(method)).toEqual(
          methodReturnDoc({
            type: { name: 'number' },
          }),
        );
      });

      test('resolves flow types on imported functions', () => {
        const def = parse.statement<ClassDeclaration>(
          `
          class Foo {
            foo = baz
          }
          import baz from 'baz';
        `,
          mockImporter,
        );
        const method = def
          .get('body')
          .get('body')[0] as NodePath<ClassProperty>;

        expect(getMethodDocumentation(method)).toEqual(
          methodReturnDoc({
            type: { name: 'number' },
          }),
        );
      });
    });

    describe('private', () => {
      test('ignores private typescript methods', () => {
        const def = parseTypescript.statement<ClassDeclaration>(
          `
          class Foo {
            private foo() {}
          }
        `,
        );
        const method = def.get('body').get('body')[0] as NodePath<ClassMethod>;

        expect(getMethodDocumentation(method)).toMatchSnapshot();
      });

      test('ignores private methods', () => {
        const def = parse.statement<ClassDeclaration>(
          `class Foo {
            #foo() {}
          }`,
          { parserOpts: { plugins: ['classPrivateMethods'] } },
        );
        const method = def
          .get('body')
          .get('body')[0] as NodePath<ClassPrivateMethod>;

        expect(getMethodDocumentation(method)).toMatchSnapshot();
      });
    });
  });
});
