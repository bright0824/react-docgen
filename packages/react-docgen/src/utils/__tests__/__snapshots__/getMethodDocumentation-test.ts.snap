// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMethodDocumentation name handles computed method name 1`] = `
{
  "docblock": null,
  "modifiers": [],
  "name": "@computed#foo",
  "params": [],
  "returns": null,
}
`;

exports[`getMethodDocumentation name ignores complex computed method name 1`] = `null`;

exports[`getMethodDocumentation parameters does not add type parameters to alias 1`] = `
{
  "docblock": null,
  "modifiers": [],
  "name": "foo",
  "params": [
    {
      "name": "bar",
      "optional": false,
      "type": {
        "alias": "Foo",
        "elements": [
          {
            "name": "T",
          },
        ],
        "name": "Foo",
        "raw": "Foo<T>",
      },
    },
  ],
  "returns": null,
}
`;

exports[`getMethodDocumentation parameters extracts flow type info 1`] = `
{
  "docblock": null,
  "modifiers": [],
  "name": "foo",
  "params": [
    {
      "name": "bar",
      "optional": false,
      "type": {
        "name": "number",
      },
    },
  ],
  "returns": null,
}
`;

exports[`getMethodDocumentation parameters extracts flow type info 2`] = `
{
  "docblock": null,
  "modifiers": [],
  "name": "foo",
  "params": [
    {
      "name": "bar",
      "optional": false,
      "type": {
        "name": "number",
      },
    },
  ],
  "returns": null,
}
`;

exports[`getMethodDocumentation parameters extracts flow type info on function assignment 1`] = `
{
  "docblock": null,
  "modifiers": [],
  "name": "foo",
  "params": [
    {
      "name": "bar",
      "optional": false,
      "type": {
        "name": "number",
      },
    },
  ],
  "returns": null,
}
`;

exports[`getMethodDocumentation parameters private ignores private methods 1`] = `null`;

exports[`getMethodDocumentation parameters private ignores private typescript methods 1`] = `null`;

exports[`getMethodDocumentation parameters resolves flow type info on imported functions 1`] = `
{
  "docblock": null,
  "modifiers": [],
  "name": "foo",
  "params": [
    {
      "name": "bar",
      "optional": false,
      "type": {
        "name": "number",
      },
    },
  ],
  "returns": null,
}
`;
